lib/semantics/check-omp-clause.h

```diff
@@ -0,0 +1,111 @@
+
+// OpenMP clause validity check for directives
+
+#ifndef FORTRAN_SEMANTICS_CHECK_OMP_CLAUSE_H_
+#define FORTRAN_SEMANTICS_CHECK_OMP_CLAUSE_H_
+
+#include "semantics.h"
+#include "../parser/parse-tree.h"
+#include <typeindex>
+
+namespace Fortran::semantics {
+
+class OmpClauseChecker : public virtual BaseChecker {
+public:
+  OmpClauseChecker(SemanticsContext &context) : context_{context} {}
+
+  void Enter(const parser::OmpBlockDirective &);
+  void Enter(const parser::OmpBlockDirective::Parallel &);
+  void Leave(const parser::OmpEndBlockDirective &);
+
+  void Enter(const parser::OmpClause &);
+  void Enter(const parser::OmpClause::Defaultmap &);
+  void Enter(const parser::OmpClause::Inbranch &);
+  void Enter(const parser::OmpClause::Mergeable &);
+  void Enter(const parser::OmpClause::Nogroup &);
+  void Enter(const parser::OmpClause::Notinbranch &);
+  void Enter(const parser::OmpClause::Untied &);
+  void Enter(const parser::OmpClause::Collapse &);
+  void Enter(const parser::OmpClause::Copyin &);
+  void Enter(const parser::OmpClause::Copyprivate &);
+  void Enter(const parser::OmpClause::Device &);
+  void Enter(const parser::OmpClause::DistSchedule &);
+  void Enter(const parser::OmpClause::Final &);
+  void Enter(const parser::OmpClause::Firstprivate &);
+  void Enter(const parser::OmpClause::From &);
+  void Enter(const parser::OmpClause::Grainsize &);
+  void Enter(const parser::OmpClause::Lastprivate &);
+  void Enter(const parser::OmpClause::NumTasks &);
+  void Enter(const parser::OmpClause::NumTeams &);
+  void Enter(const parser::OmpClause::NumThreads &);
+  void Enter(const parser::OmpClause::Ordered &);
+  void Enter(const parser::OmpClause::Priority &);
+  void Enter(const parser::OmpClause::Private &);
+  void Enter(const parser::OmpClause::Safelen &);
+  void Enter(const parser::OmpClause::Shared &);
+  void Enter(const parser::OmpClause::Simdlen &);
+  void Enter(const parser::OmpClause::ThreadLimit &);
+  void Enter(const parser::OmpClause::To &);
+  void Enter(const parser::OmpClause::Uniform &);
+  void Enter(const parser::OmpClause::UseDevicePtr &);
+
+  void Enter(const parser::OmpAlignedClause &);
+  void Enter(const parser::OmpDefaultClause &);
+  void Enter(const parser::OmpDependClause &);
+  void Enter(const parser::OmpIfClause &);
+  void Enter(const parser::OmpLinearClause &);
+  void Enter(const parser::OmpMapClause &);
+  void Enter(const parser::OmpProcBindClause &);
+  void Enter(const parser::OmpReductionClause &);
+  void Enter(const parser::OmpScheduleClause &);
+
+  void EmptyAllowed() { allowedClauses_ = {}; }
+  void SetAllowed(std::set<std::type_index> allowedClauses) {
+    allowedClauses_ = allowedClauses;
+  }
+  template<typename N> void CheckAllowed(const N &node) {
+    if (allowedClauses_.empty()) {
+      context_.Say(
+          currentClauseSource_, "Internal: allowed clauses not set"_err_en_US);
+    }
+    auto it{allowedClauses_.find(typeid(node))};
```

Comment:
It is the first time I see `typeid` used in f18 and I do not know much about it. It looks like a clever way to avoid defining an `enum`. Are you sure that it does not come with runtime over-head? Should `typeid(node)` be `typeid(N)` to avoid any ploymorphism trouble?

Reply:
I was expecting `typeid` to check the type instead of objects, so most of the work should be done during compile time. But I did see runtime overhead is mentioned in cppreference.com:

> When applied to an expression of polymorphic type, evaluation of a typeid expression may involve runtime overhead (a virtual table lookup), otherwise typeid expression is resolved at compile time.

Reply:
`typeid` depends on RTTI and that's not allowed in llvm.  Sorry.

https://llvm.org/docs/CodingStandards.html#do-not-use-rtti-or-exceptions

Reply:
Thanks, that's good to know. @klausler

Reply:
Should we be building with `-fno-rtti` and `-fno-exceptions`?

Reply:
`-fno-rtti`, definitely.  Not sure about the other, since many STL errors (like `.value()` on an empty `optional`) are reported by throwing exceptions.